File: mos/backend/app/services/funnel_template_categories.py
from __future__ import annotations


_TEMPLATE_CATEGORY_ALIASES = {
    "presales": "presales",
    "pre-sales": "presales",
    "pre_sales": "presales",
    "sales": "sales",
}


def normalize_template_category(value: str | None) -> str | None:
    cleaned = str(value or "").strip().lower()
    if not cleaned:
        return None
    return _TEMPLATE_CATEGORY_ALIASES.get(cleaned)


def infer_template_category_from_id(template_id: str | None) -> str | None:
    cleaned = str(template_id or "").strip().lower()
    if not cleaned:
        return None
    if cleaned in {"pre-sales-listicle", "pre_sales_listicle"}:
        return "presales"
    if cleaned in {"sales-pdp", "sales_pdp"}:
        return "sales"
    if cleaned.startswith(("presales-", "pre-sales-", "pre_sales_")):
        return "presales"
    if cleaned.startswith(("sales-", "sales_")):
        return "sales"
    return None


def resolve_funnel_template_category(template_id: str | None) -> str | None:
    category = normalize_template_category(template_id)
    if category:
        return category
    return infer_template_category_from_id(template_id)


def resolve_funnel_template_public_stage(template_id: str | None) -> str | None:
    category = resolve_funnel_template_category(template_id)
    if category == "presales":
        return "pre_sales"
    if category == "sales":
        return "sales"
    return None


def resolve_funnel_template_artifact_slug(template_id: str | None) -> str | None:
    category = resolve_funnel_template_category(template_id)
    if category == "presales":
        return "presales"
    if category == "sales":
        return "sales"
    return None

File: mos/frontend/src/lib/funnelTemplateCategories.ts
export type FunnelTemplateCategory = "presales" | "sales" | "compliance";

const TEMPLATE_CATEGORY_LABELS: Record<FunnelTemplateCategory, string> = {
  presales: "Pre-Sales Templates",
  sales: "Sales Templates",
  compliance: "Compliance Templates",
};

const TEMPLATE_CATEGORY_ALIASES: Record<string, FunnelTemplateCategory> = {
  presales: "presales",
  "pre-sales": "presales",
  pre_sales: "presales",
  sales: "sales",
  compliance: "compliance",
};

export function normalizeFunnelTemplateCategory(value: string | null | undefined): FunnelTemplateCategory | null {
  const cleaned = typeof value === "string" ? value.trim().toLowerCase() : "";
  if (!cleaned) {
    return null;
  }
  return TEMPLATE_CATEGORY_ALIASES[cleaned] ?? null;
}

export function resolveFunnelTemplateCategory(templateId: string | null | undefined): FunnelTemplateCategory | null {
  const cleaned = typeof templateId === "string" ? templateId.trim().toLowerCase() : "";
  if (!cleaned) {
    return null;
  }
  if (cleaned === "pre-sales-listicle" || cleaned === "pre_sales_listicle") {
    return "presales";
  }
  if (cleaned === "sales-pdp" || cleaned === "sales_pdp") {
    return "sales";
  }
  if (cleaned.startsWith("presales-") || cleaned.startsWith("pre-sales-") || cleaned.startsWith("pre_sales_")) {
    return "presales";
  }
  if (cleaned.startsWith("sales-") || cleaned.startsWith("sales_")) {
    return "sales";
  }
  if (cleaned.startsWith("compliance-") || cleaned.startsWith("compliance_")) {
    return "compliance";
  }
  return null;
}

export function funnelTemplateCategoryLabel(category: FunnelTemplateCategory | null | undefined): string | null {
  if (!category) {
    return null;
  }
  return TEMPLATE_CATEGORY_LABELS[category];
}
